### PR TITLE
tests updated

### PR DESCRIPTION
**Pull request by iambeone**
Tests update to work with different networks.
Networks configuration in tests/e2e/networks.json
Call npm run test:e2e:local -- --network=1

Thank you! 🚀

---

For contributor:

- [ ] Added changes entries. Run `yarn changelog` for a guided process.
- [ ] Reviewed `Files changed` in the github PR explorer
- [ ] Attach screenshots of the UI components on the PR description (if applicable)
- [ ] Scope of work approved for big PRs

For reviewer:

- [ ] Manually tested the changes on the UI

### CODE REVIEW

**Comment by faboweb** [comment]
Does this test against a live version /deployment preview of Lunie?

File: tests/e2e/delegation.spec.js [hunk 1]
```diff
@@ -1,138 +1,142 @@
...
-    },
-    [selector, option]
-  )
-}
+const { actionModalCheckout, nextBlock, waitForText } = require("./helpers.js")
+
+module.exports = {
+  "Delegate Action": async function(browser) {
+    // move to according page
+    browser.url(browser.launch_url + "/validators", async () => {
```
Comment (faboweb) [inline tests/e2e/delegation.spec.js]: browser.url doesn't return a promise?

File: tests/e2e/globals.js [hunk 2]
```diff
@@ -1,125 +1,174 @@
...
+
+  async before() {
+    await schemaAvailable()
+    await apiUp()
+  },
+
+  beforeEach(browser, done) {
+    let network = 1
+    let networkData = {}
+    process.argv.map(arg => {
```
Comment (faboweb) [inline tests/e2e/globals.js]: refactor into `getNetworkFromInput`?

File: tests/e2e/globals.js [hunk 3]
```diff
@@ -1,125 +1,174 @@
...
+  { address, network, wallet, password, name, stakeAmount, restakeAmount }
+) {
+  // clear parameters
+  browser.globals.availableAtoms = 0
+  browser.globals.denom = ""
+  browser.globals.password = password
+  browser.globals.stakeAmount = stakeAmount
+  browser.globals.restakeAmount = restakeAmount
+  // test if the test account was funded as we need the account to have funds in the tests
+  axios
```
Comment (faboweb) [inline tests/e2e/globals.js]: why did you not leverage `async` `await` syntax here?

File: tests/e2e/signin.spec.js [hunk 4]
```diff
@@ -1,133 +1,151 @@
...
-  browser.waitForElementVisible(`body`, 10000, true)
-  browser.waitForElementVisible(`#app-content`, 10000, true)
-  signOut(browser)
-  signIn(browser)
-
-  browser.waitForElementVisible("#session-welcome", 10000, true)
-}
+module.exports = {
+  /*
+  "Sign in with local account": async function(browser) {
```
Comment (faboweb) [inline tests/e2e/signin.spec.js]: this didn't work? insecure mode not set maybe?

File: tests/e2e/delegation.spec.js (hunk 1 above)
Comment (faboweb) [reply tests/e2e/delegation.spec.js]: I guess it doesn't

File: tests/e2e/networks.json [hunk 5]
```diff
@@ -0,0 +1,19 @@
+[
```
Comment (faboweb) [inline tests/e2e/networks.json]: I propose a structure:

```
{
  networkdID: {
    ...networkConfig
  }
}
```

so we can access the network configs directly by id

File: tests/e2e/globals.js [hunk 6]
```diff
@@ -1,125 +1,174 @@
...
+
+  beforeEach(browser, done) {
+    let network = 1
+    let networkData = {}
+    process.argv.map(arg => {
+      if (arg.indexOf("--network=") !== -1)
+        network = arg.slice(arg.indexOf("=") + 1)
+    })
+    if (networks.length >= network) {
+      networkData = networks[network - 1]
```
Comment (faboweb) [inline tests/e2e/globals.js]: how about using the networkId string instead of an intransparent number

File: tests/e2e/globals.js (hunk 3 above)
Comment (faboweb) [reply tests/e2e/globals.js]: I also think this could be refactored into a function `waitForFunds(network, address)`

File: tests/e2e/globals.js [hunk 7]
```diff
@@ -1,125 +1,174 @@
...
+      })
+      // switching to homepage
+      browser.url(browser.launch_url, () => {
+        // beforeEach done
+        done()
+      })
+    })
+}
+
+async function apiUp() {
```
Comment (faboweb) [inline tests/e2e/globals.js]: if we live test, we don't need this anymore, right?

File: tests/e2e/globals.js [hunk 8]
```diff
@@ -1,125 +1,174 @@
...
+      apiUp = true
+    } catch (err) {
+      console.log("Failed to check API", err)
+      console.log("Waiting 1s for API to be up")
+      await new Promise(resolve => setTimeout(resolve, 1000))
+    }
+  }
+}
+
+async function schemaAvailable() {
```
Comment (faboweb) [inline tests/e2e/globals.js]: if we live test, we don't need this anymore, right?

File: tests/e2e/networks.json (hunk 5 above)
Comment (faboweb) [inline tests/e2e/networks.json]: this file should probably not be checked in or people might f*** with our tests by draining the accounts.

**Comment by faboweb** [comment]
try to create individual PRs for inidivudal things. the api persitence could have been done in a separate PR

File: src/gql/apollo.js [hunk 9]
```diff
@@ -12,8 +12,47 @@ import config from "src/../config"
...
+      "Your API version differ from normal. Query ?iwouldliketochangetheapipersistentlyandiknowwhatido=false to reset"
+    )
+    return localStorage.getItem(`persistentapi`)
+  }
+  return false
+}
+
 const graphqlHost = urlParams =>
-  (urlParams.graphql ? decodeURIComponent(urlParams.graphql) : false) ||
+  checkPersistentAPI(urlParams) ||
```
Comment (faboweb) [inline src/gql/apollo.js]: needs to be replicated in `ActionManager`

File: tests/e2e/globals.js [hunk 10]
```diff
@@ -47,9 +82,132 @@ module.exports = {
...
+    [networkData]
+  )
+  // wallet info
+  browser.globals.wallet = tempAcc.value.wallet
+  // address
+  browser.globals.address = tempAcc.value.address
+  // switching to master account
+  await switchToAccount(browser, networkData)
+
+  // funding temp account
```
Comment (faboweb) [inline tests/e2e/globals.js]: all those should be refactored into functions so this is better to read

File: tests/e2e/delegation.spec.js [hunk 11]
```diff
@@ -103,7 +107,7 @@ module.exports = {
       "0"
     )
 
-    await nextBlock(browser)
+    //await nextBlock(browser)
```
Comment (faboweb) [inline tests/e2e/delegation.spec.js]: if we comment those out, should we remove them?

File: tests/e2e/delegation.spec.js [hunk 12]
```diff
@@ -82,16 +84,18 @@ module.exports = {
   },
   "Undelegate Action": async function(browser) {
     // be sure that the balance has updated, if we don't wait, the baseline (balance) shifts
-    await nextBlock(browser)
+    //await nextBlock(browser)
```
Comment (faboweb) [inline tests/e2e/delegation.spec.js]: if we comment those out, should we remove them?

File: tests/e2e/globals.js [hunk 13]
```diff
@@ -91,85 +45,192 @@ async function next(browser) {
...
+              first: validatorLIs[0].getAttribute("data-name"),
+              second: validatorLIs[1].getAttribute("data-name")
+            })
+          }
+          f()
+        })
+      },
+      [browser, networkData]
+    )
+    browser.globals.validatorOneName = validators.value.first
```
Comment (faboweb) [inline tests/e2e/globals.js]: this doesn't get deleted on a refresh?

Comment (iambeone) [reply tests/e2e/globals.js]: No, those variables are global for all tests during test sessions

File: package.json [hunk 14]
```diff
@@ -19,7 +19,7 @@
     "test:e2e": "node tests/e2e/local-run.js",
     "test:e2e:build": "cross-env VUE_APP_E2E=true vue-cli-service build --mode production",
     "test:e2e:browserstack": "node tests/e2e/browserstack.runner.js -c  tests/e2e/browserstack.conf.js",
-    "test:e2e:local": "kill-port 9515 && nightwatch --config ./tests/e2e/local-chrome.conf.js",
+    "test:e2e:local": "nightwatch --config ./tests/e2e/local-chrome.conf.js",
```
Comment (faboweb) [inline package.json]: why did you remove this?

File: tests/e2e/signin.spec.js [hunk 15]
```diff
@@ -17,43 +16,7 @@ module.exports = {
     await browser.waitForElementNotPresent(".session", 10000, true)
     openMenu(browser)
     await browser.waitForElementVisible("#sign-out", 10000, true)
-  },
-  */
-  "Create local account": async function(browser) {
```
Comment (faboweb) [inline tests/e2e/signin.spec.js]: Why did you remove this?

File: package.json (hunk 14 above)
Comment (iambeone) [reply package.json]: Sorry, my fault. kill-port always needs more than 3 minutes to be executed, so that's why i'm removing it always. Forgot to add it back

File: tests/e2e/signin.spec.js (hunk 15 above)
Comment (iambeone) [reply tests/e2e/signin.spec.js]: There is no "Sign in with local account" even with insecure flag. Don't know why

File: package.json (hunk 14 above)
Comment (faboweb) [reply package.json]: on Windows. Maybe there is a better plugin.

Comment (faboweb) [reply package.json]: https://github.com/tiaanduplessis/kill-port/pull/39 :/

File: tests/e2e/helpers.js [hunk 16]
```diff
@@ -37,6 +37,31 @@ async function waitFor(check, iterations = 10, timeout = 1000) {
   console.error("Condition was not meet in time")
   process.exit(2) // exiting here so the e2e tests actually fail. if not they pass
 }
+
+async function fundMasterAccount(browser, network, address) {
```
Comment (faboweb) [inline tests/e2e/helpers.js]: how does this scale?? will you create a funding function for every network? why not stick with the wallet strategy from before?

File: tests/e2e/globals.js [hunk 17]
```diff
@@ -274,6 +283,11 @@ async function createAccountAndFundIt(browser, done, networkData) {
   await storeAccountData(browser, networkData)
   // switching to master account
   await switchToAccount(browser, networkData)
+  // funding main account
+  console.log(browser.globals.availableAtoms)
```
Comment (faboweb) [inline tests/e2e/globals.js]: remove? or if needed add some information what this value is

File: tests/e2e/interface.spec.js [hunk 18]
```diff
@@ -0,0 +1,96 @@
+module.exports = {
+  "Validators search": async function(browser) {
+    // move to according page
+    browser.url(browser.launch_url + "/validators", async () => {
+      // wait until validators list load
+      await browser
+        .execute(function() {
+          return new Promise(resolve => {
+            let attempts = 5
+            const f = () => {
```
Comment (faboweb) [inline tests/e2e/interface.spec.js]: no abbreviations

File: tests/e2e/interface.spec.js [hunk 19]
```diff
@@ -0,0 +1,96 @@
+module.exports = {
+  "Validators search": async function(browser) {
+    // move to according page
+    browser.url(browser.launch_url + "/validators", async () => {
+      // wait until validators list load
+      await browser
+        .execute(function() {
+          return new Promise(resolve => {
```
Comment (faboweb) [inline tests/e2e/interface.spec.js]: use async await to avoid nesting hell

File: tests/e2e/interface.spec.js [hunk 20]
```diff
@@ -0,0 +1,96 @@
+module.exports = {
```
Comment (faboweb) [inline tests/e2e/interface.spec.js]: please refactor this file. this is really hard to read

File: tests/e2e/signin.spec.js [hunk 21]
```diff
@@ -104,22 +114,60 @@ function signIn(browser) {
...
+          JSON.stringify({
+            name: `rich_account`,
+            address: `cosmos1ek9cd8ewgxg9w5xllq9um0uf4aaxaruvcw4v9e`,
+            wallet: `ae1d20a49e1085cca29a71e270c6f64f8f86794cb67c6922caea6bcba0ed9e60g+nSTgP8/wHpWaomDkhW/7g2Xldvno3VRFggvdpWIDrRV+n4BJtpk3UpLKo0K3SDL5dRzxz3NmGFnSA8znggFmtesdqu6jWJYzSNqaQhM/gCPTVabF7t1UHaybze1NRlYcm/wl5oOyXRpki6ugOHxNhF7+4wlzhYxMilAB7ekDB4+VVHoPMUinU4dsUdtC4XwDUA0rbX1TTmrh+i1eBp6UTQ+nHGiZXL1TkhhR1mE0fR3bLRunz5XagYtjoST33pecQWzqeaZZQ/mgm9QXu/i+ymfbnPQkh8ivx+J6/d2RfZuAV4NnwFZDUr7CzPX4TU`
+          })
+        )
+        return true
+      }, [])
+      await browser.refresh()
+      openMenu(browser).then(async () => {
```
Comment (faboweb) [inline tests/e2e/signin.spec.js]: can't await here?

File: tests/e2e/signin.spec.js [hunk 22]
```diff
@@ -104,22 +114,60 @@ function signIn(browser) {
...
+        // check if we are already singed in
+        const isSigned = await browser.execute(function() {
+          return new Promise(resolve => {
+            let attempts = 3
+            const f = () => {
```
Comment (faboweb) [inline tests/e2e/signin.spec.js]: can't you make this more linear? like in a simple async while loop?

File: tests/e2e/signin.spec.js [hunk 23]
```diff
@@ -104,22 +114,60 @@ function signIn(browser) {
...
+              const signOutElement = document.querySelector(
+                ".user-box-address #sign-out"
+              )
+              const singInElement = document.querySelector(".app-menu #sign-in")
+              if (!signOutElement < 2 && attempts-- > 0) {
```
Comment (faboweb) [inline tests/e2e/signin.spec.js]: why are there 2 #sign-in elements?

**Comment by faboweb** [comment]
There are conflicts on here and the unit tests are failing

File: tests/e2e/networks.json (hunk 5 above)
Comment (faboweb) [reply tests/e2e/networks.json]: ok, we can leave the wallets in for now. I think it is cleaner though if we remove the whole file and add it via env variable later

File: tests/e2e/helpers.js (hunk 16 above)
Comment (iambeone) [reply tests/e2e/helpers.js]: Yes, I suppose to or it could just notify us somehow. This function only trigger when master account has a low balance. Newly created accounts are still funded from master, this function is only for funding master account

File: tests/e2e/globals.js (hunk 17 above)
Comment (iambeone) [reply tests/e2e/globals.js]: Sorry, this needs to be removed

File: tests/e2e/interface.spec.js (hunk 18 above)
Comment (iambeone) [reply tests/e2e/interface.spec.js]: You mean f()? But this is a common name for a function that will be executed only in this block. Any suggestions about the name?

File: tests/e2e/signin.spec.js (hunk 23 above)
Comment (iambeone) [reply tests/e2e/signin.spec.js]: That is sign-it and sign-out. For some reasons, we can't just rely on getting sign-in element cause it's always showing before it will be replaced with sign-out. So we need before sign-out will show (or not)

**Comment by Bitcoinera** [comment]
> There are conflicts on here and the unit tests are failing

Strangely I have seen no conflicts when merging and unit tests are fine in local :thinking:

File: tests/e2e/helpers.js (hunk 16 above)
Comment (faboweb) [reply tests/e2e/helpers.js]: this only works for gaia testnet is what I meant. so this should if we want to keep this be refactored into some structure that tells us this only handles gaia testnet.

File: tests/e2e/interface.spec.js (hunk 18 above)
Comment (faboweb) [reply tests/e2e/interface.spec.js]: see my refactoring removing the need for this. We have a code guideline to never use abbreviations. Always describe what this does. If you can't describe what this does, this probably needs to be refactored.

File: tests/e2e/signin.spec.js (hunk 23 above)
Comment (faboweb) [reply tests/e2e/signin.spec.js]: But we can rely on the sign-out element as it only shows when the user is signed in, right?

**Comment by faboweb** [comment]
changelog is missing
